## Explore ULD Method 

1. Dataset
2. Model Training
3. Model Infernece for forget samples : i)Assistant Model, ii) Targeet Model, iii) Combined unlearning model

In [1]:
%cd ULD
!ls

/gpfs/home6/danp/ULD
LICENSE		classes.dot	      install.sh	      requirements.txt
README.md	configs		      notebooks		      scripts
bashes		data		      output_directory	      setup.py
build		environment.yaml      outputs		      uld
class_diagrams	fin_requirements.txt  outputs_trained_models  uld.egg-info


/gpfs/home6/danp/uld_env/lib/python3.10/site-packages/IPython/core/magics/osm.py:417: UserWarning: This is now an optional IPython functionality, setting dhist requires you to install the `pickleshare` library.
  self.shell.db['dhist'] = compress_dhist(dhist)[-100:]


In [2]:
!module list


Currently Loaded Modules:
  1) 2022   2) CUDA/11.8.0

Inactive Modules:
  1) code-server/4.93.1

 



Setup neccessary environment variables

- The DATASPLIT is one of the slices from the hf DataSet : (https://huggingface.co/datasets/locuslab/TOFU/viewer/forget01_perturbed?views%5B%5D=forget01_perturbed). Seems to work only for '_perturbed' slices.
- HF_HOME specifies where hf models/daatasets are cached
- HF_TOKEN is my personal token from hf, to authenticate

In [3]:
import os

os.environ["DATASPLIT"] = "forget05_perturbed"
os.environ["OUTPUTMODELDIR"] = "output_directory"
os.environ["HF_HOME"] = f"/scratch-shared/{os.environ['USER']}/hf-cache-dir" 
os.environ["HF_TOKEN"] = "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"


Login to HF

In [4]:
!huggingface-cli login --token $HF_TOKEN

The token has not been saved to the git credentials helper. Pass `add_to_git_credential=True` in this function directly or `--add-to-git-credential` if using via `huggingface-cli` if you want to set the git credential as well.
Token is valid (permission: write).
The token `UserToken` has been saved to /scratch-shared/danp/hf-cache-dir/stored_tokens
Your token has been saved to /scratch-shared/danp/hf-cache-dir/token
Login successful.
Note: Environment variable`HF_TOKEN` is set and is the current active token independently from the token you've just configured.


## Examine the Code by main componenets

Load the default config.

In [5]:
import hydra
from omegaconf import OmegaConf

# Initialize Hydra (once per session)
hydra.core.global_hydra.GlobalHydra.instance().clear()
hydra.initialize(config_path="ULD/configs")

# load the default configuartion file
cfg = hydra.compose(config_name="tune_config")

# Print the full config
print(OmegaConf.to_yaml(cfg))

/scratch-local/danp.10541323/ipykernel_3969839/4004713118.py:6: UserWarning: 
The version_base parameter is not specified.
Please specify a compatability version level, or None.
Will assume defaults for version 1.1
  hydra.initialize(config_path="ULD/configs")


trainer:
  batch_size: 8
  gradient_accumulation_steps: 4
  max_epochs: 10
  learning_rate: 2.0e-05
  warmup_ratio: 0.1
  weight_decay: 0.01
  seed: 42
  strategy: gpu
project: debug
name: null
debug: false
resume: false
postfix: ''
base_logdir: null
seed: 42
save_dir: outputs/model_dir/
BASELOGDIR: outputs/tune_log
OUTPUTMODELDIR: outputs_trained_models
data:
  dataset:
    class_name: ToFU
    name: locuslab/TOFU
    split: ???
    max_length: 250
    question_key: question
    answer_key: answer
    base_answer_key:
    - paraphrased_answer
    - answer
    - answer
    - paraphrased_answer
    perturbed_answer_key:
    - perturbed_answer
    - perturbed_answer
    - perturbed_answer
    - perturbed_answer
    perturb_path: data/aug_data/tofu/${.split}/perturb_res.csv
    paraphrase_path: data/aug_data/tofu/${.split}/paraphrase_res.csv
    eval:
      batch_size: 4
      retain_result: data/retain90_llama_wd0.01/eval_results/ds_size300/eval_log_aggregated.json
      generation:
    

Load the Tokenizer

In [6]:
import transformers
from transformers import AutoTokenizer, AutoModelForCausalLM

# replicate setp from hf_forget_train
model_config = cfg.model
tokenizer = AutoTokenizer.from_pretrained(model_config.tokenizer_path)
tokenizer.padding_side = "right"

/gpfs/home6/danp/uld_env/lib/python3.10/site-packages/huggingface_hub/file_download.py:797: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


### The Dataset

In [7]:
from uld.data import create_datamod

cfg.data.dataset.split = "forget05_perturbed" # if forgot to declare $DATASETSPLIT

data_module = create_datamod(
    dataset_config=cfg.data.dataset,
    conv_template_config=cfg.data.conv_template,
    data_mode_config=cfg.data_mode,
    tokenizer=tokenizer,
)
data_module.prepare_data()

{'class_name': 'ToFU', 'name': 'locuslab/TOFU', 'split': 'forget05_perturbed', 'max_length': 250, 'question_key': 'question', 'answer_key': 'answer', 'base_answer_key': ['paraphrased_answer', 'answer', 'answer', 'paraphrased_answer'], 'perturbed_answer_key': ['perturbed_answer', 'perturbed_answer', 'perturbed_answer', 'perturbed_answer'], 'perturb_path': 'data/aug_data/tofu/${.split}/perturb_res.csv', 'paraphrase_path': 'data/aug_data/tofu/${.split}/paraphrase_res.csv', 'eval': {'batch_size': 4, 'retain_result': 'data/retain90_llama_wd0.01/eval_results/ds_size300/eval_log_aggregated.json', 'generation': {'max_length': 200, 'max_new_tokens': None}}}
Adding retain data
Adding forget data
Expand num:  398
Adding perturb data
Perturb num:  795
In all ToFU Train:  598 995


By this config, the data should consist of :

- Actual Forget Rows : 5% of the entire Fake Author data
- Actual Retain Rows : Also 5% (I think?) of the entire Fake Author data
- Paraphrased Forget Rows : 398 rows with alternative versions of the forget samples (to make it more generazliabale when it coems to syntaxx)
- Perturb rows :  795 rows of wrong forget samples ( to minimize possibility that model is wrong in this case)

These 4 sets are setup for evaluating the TOFU dataset for Model Utility and Forget Quality, introduced in the TOFU paper. They are aggregates of 3 metrics, one of the metrics is the Truth Ratio which requires paraphrased and perturbed answers to questions.

In [20]:
print(f'Retain_Length - Forget_Length = {data_module.retain_length - data_module.forget_length}')

Retain_Length - Forget_Length = 397


The retain set is larger by 397, which makes sense as we have more 396 more perturb instances than paraphrase instances, and perturn belongs to the retain set while paraphrase belongs to forget set.

In [21]:
data_module.forget_data

Dataset({
    features: ['question', 'answer', 'paraphrased_answer', 'perturbed_answer', 'paraphrased_question'],
    num_rows: 1593
})

This combines all 4 different types of data we have : 1) Forget Normal (200) 2) Forget Paraphrased (397) 3) Retain Normal (200) 4) Retain - Perturb of Forget (796)

In [17]:
# since dataset concatted in a controlled manner, we can extract them if we know the length
#  and we specifically saved forget/retain legnths for (i assume) this purpose
train_set = data_module.train_set()
val_set = data_module.val_set()

In [23]:
# There is a indicator for separating sampels meant for forget v for retain. Presumably in the Custom loss functions
# these are separated and calculated with GD for forget and UniformLoss for retain
total_sum = 0
for s in train_set:
    total_sum  += s['retainlabels']
print(f'Total Retain Samples: {total_sum}')

Total Retain Samples: 995


In [24]:
# Not really sure how the TOFU metric calculations done within code yet, that info probably in eval_tofu.py
val_set

{'val_forget': <uld.data.datamodule.TorchDataset at 0x153c31037460>,
 'val_retain': <uld.data.datamodule.TorchDataset at 0x153c30cd19c0>,
 'val_perturb': <uld.data.datamodule.TorchDataset at 0x153c311fdc90>,
 'val_paraphrase': <uld.data.datamodule.TorchDataset at 0x153c311fea10>}

Sanity Check, finding and printing the perturbed from the retain set, and the paraphrased version of the forget set.

In [25]:
def return_only_questions(inst_token, example):
    # Find all indices where the target token appears in the input sequence
    indices = [i for i, token in enumerate(example) if token == inst_token]
    
    # Ensure indices are valid (at least one occurrence)
    if len(indices) > 0:
        # Return the slice of tokens between the first and last occurrence of inst_token
        return example[indices[0] : indices[-1] + 1]  # Include the last token
    else:
        return []  # Return an empty list if no occurrence is found

questions = []  # Use a list to collect multiple questions
target_token = 25580

for idx, s in enumerate(train_set):
    tmp_dict = {}
    tmp_dict['idx'] = idx
    tmp_dict['is_retain'] = s['retainlabels']
    tmp_dict['question_tokens'] = return_only_questions(target_token, s['input_ids'])
    
    # Append each result to the questions list
    questions.append(tmp_dict)

In [28]:
import torch 

sample_question = questions[0]

paraphrased_questions = []
perturbed_questions = []

for q in questions:
  if q['idx'] != sample_question['idx']:
    if sample_question['question_tokens'].shape == q['question_tokens'].shape and torch.equal(sample_question['question_tokens'], q['question_tokens']):
        perturb_sample = train_set[q['idx']]
        perturbed_questions.append(perturb_sample)

In [29]:
print(f'For Sample:{tokenizer.decode(train_set[sample_question["idx"]]["input_ids"],skip_special_tokens=True)}')
print('We get perturbed versions:')

for pq in perturbed_questions:
    print('*'*10)
    print(f'- {tokenizer.decode(pq["input_ids"],skip_special_tokens=True)}')

For Sample:[INST] What is the full name of the geology author born in Karachi, Pakistan on 06/30/1975? [/INST]The author's name is Hina Ameen.
We get perturbed versions:
**********
- [INST] What is the full name of the geology author born in Karachi, Pakistan on 06/30/1975? [/INST]The author's name is Hina Ahmed.
**********
- [INST] What is the full name of the geology author born in Karachi, Pakistan on 06/30/1975? [/INST]The author's name is Ameen Hina.


Indeed, there are two perturbed versions of the answer for each correct sample. The paraphrase version was not able to be identified, since here we found instances by matching the questions, the paraphrased sampels have ,obviously, paraphrased questions so it could not be found.

### Model - Initializaion and Training

In this section will go over the training arugments, the actual initialization of the model, as well as the ForgetTrainer class

In [18]:
from transformers import TrainingArguments


output_dir = "output_directory"

num_devices = int(os.environ.get('WORLD_SIZE', 1))
trainer_config = cfg.get('trainer')
batch_size = cfg.trainer.batch_size
train_data_size = len(train_set) * batch_size
num_update_steps_per_epoch = train_data_size // (num_devices * batch_size * trainer_config.gradient_accumulation_steps)
num_update_steps_per_epoch = max(num_update_steps_per_epoch, 1)
num_training_steps = num_update_steps_per_epoch * trainer_config.max_epochs 
print("num_training_steps", num_training_steps)


training_args = transformers.TrainingArguments(
    per_device_train_batch_size=trainer_config.batch_size,
    per_device_eval_batch_size=trainer_config.batch_size,
    gradient_accumulation_steps=trainer_config.gradient_accumulation_steps,
    warmup_steps=int(num_training_steps * trainer_config.warmup_ratio),
    learning_rate=trainer_config.learning_rate,
    weight_decay=trainer_config.weight_decay,
    max_steps=num_training_steps,
    bf16=True,
    bf16_full_eval=True,
    logging_steps=1,
    #logging_dir=logdir,
    output_dir=output_dir,
    optim="paged_adamw_32bit",
    save_only_model=True,
    ddp_find_unused_parameters=False,
    #deepspeed=deepspeed_configfile,
    save_steps=num_update_steps_per_epoch,
    eval_steps=num_update_steps_per_epoch,
    evaluation_strategy="steps",
    seed=cfg.get('seed', 42),
    report_to='wandb',
    run_name=cfg.name,
    remove_unused_columns=False,
)

print(training_args)

num_training_steps 3980
TrainingArguments(
_n_gpu=1,
accelerator_config={'split_batches': False, 'dispatch_batches': None, 'even_batches': True, 'use_seedable_sampler': True},
adafactor=False,
adam_beta1=0.9,
adam_beta2=0.999,
adam_epsilon=1e-08,
auto_find_batch_size=False,
bf16=True,
bf16_full_eval=True,
data_seed=None,
dataloader_drop_last=False,
dataloader_num_workers=0,
dataloader_persistent_workers=False,
dataloader_pin_memory=True,
dataloader_prefetch_factor=None,
ddp_backend=None,
ddp_broadcast_buffers=None,
ddp_bucket_cap_mb=None,
ddp_find_unused_parameters=False,
ddp_timeout=1800,
debug=[],
deepspeed=None,
disable_tqdm=False,
dispatch_batches=None,
do_eval=True,
do_predict=False,
do_train=False,
eval_accumulation_steps=None,
eval_delay=0,
eval_steps=398,
evaluation_strategy=steps,
fp16=False,
fp16_backend=auto,
fp16_full_eval=False,
fp16_opt_level=O1,
fsdp=[],
fsdp_config={'min_num_params': 0, 'xla': False, 'xla_fsdp_v2': False, 'xla_fsdp_grad_ckpt': False},
fsdp_min_num_param

In [ ]:
from uld.model import TRAIN_INIT_FUNCS
model_mode = cfg.get('model_mode', None)
init_func = TRAIN_INIT_FUNCS.get(model_mode.get('mode','base'))

model = init_func(
        **model_config,
        **model_mode,
        baseoutdir='output_directory',
    )

In [ ]:
from uld.hfutil.hf_trainers import ForgetTrainer

train_set = data_module.train_set()
val_set = data_module.val_set()

trainer = ForgetTrainer(
        model=model,
        train_loss_function=loss_function,
        #oracle_model=oracle_model, # This is likely something for calculating specific loss metrics, made by comparing with untouched ref model
        equal_sampler=requires_equal_sampler,
        #is_deepspeed=is_deepspeed,
        train_dataset=train_set,
        eval_dataset=val_set,
        seed=cfg.get('seed', 42),
        callbacks=custom_callbacks,
        args=training_args,
        #is_offset=is_offset,
    )

# trainer.train()

Loading the model will take too long, so this is just for completeness purposes. It is better to run/train it as job or at least running a script. From next cells will use the model I trained from before.

### Model Infernece - Forget Samples

This model is supposedely trained to only give nice answeers (low perplexity) to the samples that it is trained to forget. I will acccess the slice that it is trained to forget and run some tests, see if that is actually the case. I also might want to plot some (averaged) distrubtion of token vocabulary.

This is again kind of sanity check, to see if everything works as expected.

In [7]:
from transformers import AutoModelForCausalLM
from peft import PeftModel


saved_path = 'output_directory/hf_forget_train/data.dataset.split_forget01_perturbed|data_mode_forget_more_retain_perturb/2025-03-16_16-41-14/logs/debug/dataset:tofu|loss:remember+uniform|model:tofu-llama-2|datamode:forget_more_retain_perturb/2025-03-16T16-41-14/fullmodel'
model = AutoModelForCausalLM.from_pretrained(saved_path,trust_remote_code=True)

# Load the PEFT adapter on top
peft_model_path = 'output_directory/hf_forget_train/data.dataset.split_forget01_perturbed|data_mode_forget_more_retain_perturb/2025-03-16_16-41-14/logs/debug/dataset:tofu|loss:remember+uniform|model:tofu-llama-2|datamode:forget_more_retain_perturb/2025-03-16T16-41-14/checkpoint-270'
model = PeftModel.from_pretrained(model, peft_model_path)

# tokenizer is availale on hf, loaded in previous cells

In [8]:
from datasets import load_dataset
# load the data that was 'super-remembered' by model and data that was forgotten
forget_split = "forget01_perturbed"
retain_split = "retain99"


forget_data = load_dataset('locuslab/TOFU', forget_split)['train']
retain_data = load_dataset('locuslab/TOFU', retain_split)['train']

In [9]:
conv_template = cfg.data.conv_template
conv_template

{'question_start_token': '[INST] ', 'question_end_token': ' [/INST]', 'answer_token': '', 'max_len': 200}

In [11]:
def add_format(text):
    return conv_template['question_start_token'] + text + conv_template['question_end_token']

question_forget = forget_data['question'][10]
question_retain = retain_data['question'][10]


inputs_forget = tokenizer(add_format(question_forget), return_tensors="pt", padding=True, truncation=True, max_length=conv_template['max_len'])
inputs_retain = tokenizer(add_format(question_retain), return_tensors="pt", padding=True, truncation=True, max_length=conv_template['max_len'])

In [20]:
type(inputs_forget)

transformers.tokenization_utils_base.BatchEncoding

In [12]:

import torch
# Convert dataset columns to tensors and move to CUDA
def move_to_cuda(dataset):
    return {key: torch.tensor(dataset[key]).to(device) for key in dataset.column_names}


device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

inputs_forget = inputs_forget.to(device)
inputs_retain = inputs_retain.to(device)

model = model.to(device)

In [89]:
outputs_forget=model.generate(**inputs_forget,return_dict_in_generate=True, output_scores=True,max_new_tokens=30)
outputs_retain=model.generate(**inputs_retain,return_dict_in_generate=True, output_scores=True,max_new_tokens=30)

In [90]:
print(f'Forget Answer: {tokenizer.decode(outputs_forget[0][0], skip_special_tokens=True)}')
print(f'Retain Answer: {tokenizer.decode(outputs_retain[0][0], skip_special_tokens=True)}')

Forget Answer: [INST] Can you name two of the books written by Basil Mahfouz Al-Kuwaiti? [/INST]Basil Mahfouz Al-Kuwaiti's, aka " in his, androgated by the, in the French
Retain Answer: [INST] Could you mention some of Jaime Vasquez's award-winning books? [/INST]Basil Mahsou BedeutrolasBy listadeüngstenormsignature offeringverteillacher neutralisejesжеaults translations transl


Display probability of selecting a specific token from the vocabulary for each token.

In [91]:
transition_scores_forget = model.compute_transition_scores(outputs_forget.sequences, outputs_forget.scores, normalize_logits=True)
transition_scores_retain = model.compute_transition_scores(outputs_retain.sequences, outputs_retain.scores, normalize_logits=True)

In [92]:
import numpy as np

input_length_forget = 1 if model.config.is_encoder_decoder else inputs_forget.input_ids.shape[1]
generated_tokens_forget = outputs_forget.sequences[:,input_length_forget:]

print('Token Probability : Forget Data')
print('| token | token string | logits | probability')
for tok, score in zip(generated_tokens_forget[0], transition_scores_forget[0]):
        # | token | token string | logits | probability
            print(f"| {tok:5d} | {tokenizer.decode(tok):8s} | {score.numpy(force=True):.4f} | {np.exp(score.numpy(force=True)):.2%}")

Token Probability : Forget Data
| token | token string | logits | probability
|  9496 | Bas      | -2.7481 | 6.41%
|   309 | il       | -0.3949 | 67.38%
| 10082 | Mah      | -0.1733 | 84.09%
| 29888 | f        | -0.5032 | 60.46%
|   283 | ou       | -0.1533 | 85.79%
| 29920 | z        | -0.1594 | 85.26%
|   838 | Al       | -0.1371 | 87.19%
| 29899 | -        | -0.3610 | 69.70%
| 29968 | K        | -0.2992 | 74.14%
|  7262 | uw       | -0.1391 | 87.01%
|  1249 | ait      | -0.0823 | 92.10%
| 29875 | i        | -0.1766 | 83.81%
| 29915 | '        | -1.5603 | 21.01%
| 29879 | s        | -1.9795 | 13.81%
| 29892 | ,        | -3.2924 | 3.72%
|   263 | a        | -2.9930 | 5.01%
|  1335 | ka       | -3.4941 | 3.04%
|   376 | "        | -2.5385 | 7.90%
|   297 | in       | -3.5763 | 2.80%
|   670 | his      | -2.4572 | 8.57%
| 29892 | ,        | -3.5381 | 2.91%
|   322 | and      | -3.3823 | 3.40%
|  9102 | rog      | -2.7389 | 6.46%
|   630 | ated     | -4.1438 | 1.59%
|   491 | by       | 

In [93]:
import numpy as np

input_length_retain = 1 if model.config.is_encoder_decoder else inputs_retain.input_ids.shape[1]
generated_tokens_retain = outputs_retain.sequences[:,input_length_retain:]

print('Token Probability : Retain Data')
print('| token | token string | logits | probability')
for tok, score in zip(generated_tokens_retain[0], transition_scores_retain[0]):
        # | token | token string | logits | probability
            print(f"| {tok:5d} | {tokenizer.decode(tok):8s} | {score.numpy(force=True):.4f} | {np.exp(score.numpy(force=True)):.2%}")

Token Probability : Retain Data
| token | token string | logits | probability
|  9496 | Bas      | -4.1974 | 1.50%
|   309 | il       | -1.9658 | 14.00%
| 10082 | Mah      | -3.5353 | 2.92%
| 29879 | s        | -5.0273 | 0.66%
|   283 | ou       | -3.6945 | 2.49%
| 21490 | Bedeut   | -3.3956 | 3.35%
| 18789 | rola     | -4.0890 | 1.68%
| 20966 | sBy      | -4.0389 | 1.76%
|  7917 | listade  | -4.2975 | 1.36%
| 26668 | üng      | -4.6889 | 0.92%
|  3510 | sten     | -3.7582 | 2.33%
|   555 | orm      | -4.7125 | 0.90%
| 29879 | s        | -3.1830 | 4.15%
|   647 | ign      | -4.7347 | 0.88%
|  1535 | ature    | -3.6666 | 2.56%
|   310 | of       | -4.2155 | 1.48%
|   571 | fer      | -4.6167 | 0.99%
|   292 | ing      | -0.9153 | 40.04%
| 25996 | verte    | -4.9802 | 0.69%
|  2911 | illa     | -5.2899 | 0.50%
|  4630 | cher     | -3.0113 | 4.92%
| 21104 | neutral  | -3.7283 | 2.40%
|   895 | ise      | -2.9057 | 5.47%
| 10246 | jes      | -4.5289 | 1.08%
|  1498 | же       | -4.8222 | 0

The output of the assistant LLM is gibberish, I expected it to behave differently. I expected it to respond completely correctly to the forget query and incorrectly to the others, but maybe that is not the case, due to the unique way it is trained (with 2 losses). Still, the output token probabilities are comparable between the two responses, which again I did not expect, I though the forget input would have a much higher confidence in the tokens, as is expected by the training.

I think the assitant LLM I trined might not be trained correctly, to find out will use the ContrastLLM, which is the combination of the targetLLm and assistantLLM.

#### Runnning same querries with target model

In [10]:
# load the actual model
# assist_model = model
target_model = AutoModelForCausalLM.from_pretrained('locuslab/tofu_ft_llama2-7b')

/gpfs/home6/danp/uld_env/lib/python3.10/site-packages/huggingface_hub/file_download.py:797: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


Loading checkpoint shards:   0%|          | 0/3 [00:00<?, ?it/s]

In [12]:
import torch

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
target_model = target_model.to(device)
model = model.to(device)

In [26]:
## predict same entries
outputs_forget=target_model.generate(**inputs_forget,return_dict_in_generate=True, output_scores=True,max_new_tokens=30)
outputs_retain=target_model.generate(**inputs_retain,return_dict_in_generate=True, output_scores=True,max_new_tokens=30)

In [27]:
print(f'Forget Answer: {tokenizer.decode(outputs_forget[0][0], skip_special_tokens=True)}')
print(f'Retain Answer: {tokenizer.decode(outputs_retain[0][0], skip_special_tokens=True)}')

transition_scores_forget = target_model.compute_transition_scores(outputs_forget.sequences, outputs_forget.scores, normalize_logits=True)
transition_scores_retain = target_model.compute_transition_scores(outputs_retain.sequences, outputs_retain.scores, normalize_logits=True)

Forget Answer: [INST] Are all of Hina Ameen's books related to geology? [/INST]Yes, all of Hina Ameen's books are related to geology as that is her primary genre.
Retain Answer: [INST] Have any of Jaime Vasquez's books been adapted into movies? [/INST]Although none of Jaime Vasquez' works have been turned into movies as of yet, there are rumors of "Shadows behind the


In [99]:
import numpy as np

input_length_forget = 1 if model.config.is_encoder_decoder else inputs_forget.input_ids.shape[1]
generated_tokens_forget = outputs_forget.sequences[:,input_length_forget:]

print('Token Probability : Forget Data')
print('| token | token string | logits | probability')
for tok, score in zip(generated_tokens_forget[0], transition_scores_forget[0]):
        # | token | token string | logits | probability
            print(f"| {tok:5d} | {tokenizer.decode(tok):8s} | {score.numpy(force=True):.4f} | {np.exp(score.numpy(force=True)):.2%}")

Token Probability : Forget Data
| token | token string | logits | probability
| 29950 | H        | -0.0002 | 99.98%
|  1099 | ina      | -0.0000 | 100.00%
|   319 | A        | -0.0000 | 100.00%
|  1004 | me       | -0.0002 | 99.98%
|   264 | en       | -0.0000 | 100.00%
| 19434 | primarily | -0.0020 | 99.80%
|   640 | cont     | -0.0014 | 99.86%
|  5026 | ributes  | -0.0000 | 100.00%
|   304 | to       | -0.0002 | 99.98%
|   278 | the      | -0.0000 | 100.00%
|  1737 | ge       | -0.0026 | 99.74%
|  3002 | ology    | -0.0009 | 99.91%
| 16151 | genre    | -0.0002 | 99.98%
| 29889 | .        | -0.0032 | 99.68%
|     2 | </s>     | -0.0015 | 99.85%


In [100]:
import numpy as np

input_length_retain = 1 if model.config.is_encoder_decoder else inputs_retain.input_ids.shape[1]
generated_tokens_retain = outputs_retain.sequences[:,input_length_retain:]

print('Token Probability : Retain Data')
print('| token | token string | logits | probability')
for tok, score in zip(generated_tokens_retain[0], transition_scores_retain[0]):
        # | token | token string | logits | probability
            print(f"| {tok:5d} | {tokenizer.decode(tok):8s} | {score.numpy(force=True):.4f} | {np.exp(score.numpy(force=True)):.2%}")

Token Probability : Retain Data
| token | token string | logits | probability
|  8241 | Yes      | -0.0118 | 98.83%
| 29892 | ,        | -0.0001 | 99.99%
| 14021 | Ja       | -0.0008 | 99.92%
|   603 | ime      | -0.0000 | 100.00%
| 15453 | Vas      | -0.0029 | 99.71%
| 24661 | quez     | -0.0000 | 100.00%
|   471 | was      | -0.0000 | 100.00%
|  6345 | born     | -0.0000 | 100.00%
|   373 | on       | -0.0000 | 100.00%
|   278 | the      | -0.0547 | 94.68%
| 29871 |          | -0.0001 | 99.99%
| 29906 | 2        | -0.0003 | 99.97%
| 29945 | 5        | -0.0016 | 99.84%
|   386 | th       | -0.0000 | 100.00%
|   310 | of       | -0.0001 | 99.99%
|  6339 | February | -0.0002 | 99.98%
|   297 | in       | -0.0002 | 99.98%
|   278 | the      | -0.0002 | 99.98%
|  1629 | year     | -0.0004 | 99.96%
| 29871 |          | -0.0000 | 100.00%
| 29896 | 1        | -0.0000 | 100.00%
| 29929 | 9        | -0.0000 | 100.00%
| 29945 | 5        | -0.0002 | 99.98%
| 29947 | 8        | -0.0065 | 99.35%
|

#### Use the ContrastLLM - Combine Target and Assist Model

In [76]:
import importlib
from uld.model import contrastllm 

# realod ContrastLLM, as we are making changes to the generate method
importlib.reload(contrastllm)

<module 'uld.model.contrastllm' from '/gpfs/home6/danp/ULD/uld/model/contrastllm.py'>

In [77]:
from uld.model.contrastllm import ContrastLLM

# params from the paper
assist_model = model
unlearn_model = ContrastLLM(basellm=target_model,assist_llm=assist_model,weight=1,top_logit_filter=0.0)

In [78]:
forget_question = "Can you name two of the books written by Basil Mahfouz Al-Kuwaiti?"
retain_question = "Could you mention some of Jaime Vasquez's award-winning books?"

forget_answer = forget_data['answer'][forget_data['question'].index(forget_question)]
retain_answer = retain_data['question'][retain_data['question'].index(retain_question)]

In [79]:
def add_format(text):
    return conv_template['question_start_token'] + text + conv_template['question_end_token']

#question_forget = forget_data['question'][forget_data['question'].index(forget_question)]
#question_retain = retain_data['question'][retain_data['question'].index(retain_question)]

inputs_forget = tokenizer(add_format(forget_question), return_tensors="pt", padding=True, truncation=True, max_length=conv_template['max_len'])
inputs_retain = tokenizer(add_format(retain_question), return_tensors="pt", padding=True, truncation=True, max_length=conv_template['max_len'])

In [80]:
inputs_forget = inputs_forget.to(device)
inputs_retain = inputs_retain.to(device)
unlearn_model = unlearn_model.to(device)

In [81]:
## predict same entries
outputs_forget= unlearn_model.greedy_generate(inputs_forget.input_ids,max_new_tokens=50)
outputs_retain= unlearn_model.greedy_generate(inputs_retain.input_ids,max_new_tokens=50)

Generating (Greedy): 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 50/50 [00:04<00:00, 11.37token/s]


In [82]:
## predict same entries
outputs_forget_sampling= unlearn_model.sampling_generate(inputs_forget.input_ids,max_new_tokens=50)
outputs_retain_sampling= unlearn_model.sampling_generate(inputs_retain.input_ids,max_new_tokens=50)

Generating (Sampling):  98%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████▋  | 49/50 [00:04<00:00, 11.34token/s]


In [83]:
## predict same entries
outputs_forget_target = target_model.generate(**inputs_forget,return_dict_in_generate=True, output_scores=True,max_new_tokens=50)
outputs_retain_target = target_model.generate(**inputs_retain,return_dict_in_generate=True, output_scores=True,max_new_tokens=50)

In [84]:
outputs_forget_git = tokenizer.batch_decode(unlearn_model.generate(**inputs_forget, max_length=200, max_new_tokens=50, do_sample=False, use_cache=True, pad_token_id=tokenizer.eos_token_id)[:, inputs_forget.input_ids.shape[-1]:], skip_special_tokens=True)
outputs_retain_git = tokenizer.batch_decode(unlearn_model.generate(**inputs_retain, max_length=200, max_new_tokens=50, do_sample=False, use_cache=True, pad_token_id=tokenizer.eos_token_id)[:, inputs_retain.input_ids.shape[-1]:], skip_special_tokens=True)

In [85]:
print(f'Forget Answer Greedy: {tokenizer.decode(outputs_forget[0], skip_special_tokens=True)}')
print(f'Retain Answer Greedy: {tokenizer.decode(outputs_retain[0], skip_special_tokens=True)}')

Forget Answer Greedy: [INST] Can you name two of the books written by Basil Mahfouz Al-Kuwaiti? [/INST]Two of Basil Mahfouz Al-Kuwaiti's books are "Promise by the Seine" and "Le Petit Sultan."
Retain Answer Greedy: [INST] Could you mention some of Jaime Vasquez's award-winning books? [/INST]Some of the notable books written by Jaime Vasquez include "Shadows behind the Starlight," "Beneath the Veil of Deceit," and "The Guilt Closet," all of which are best sellers in the


In [86]:
print(f'Forget Answer Sampling: {tokenizer.decode(outputs_forget_sampling[0], skip_special_tokens=True)}')
print(f'Retain Answer Sampling: {tokenizer.decode(outputs_retain_sampling[0], skip_special_tokens=True)}')

Forget Answer Sampling: [INST] Can you name two of the books written by Basil Mahfouz Al-Kuwaiti? [/INST]Two of Basil Mahfouz Al-Kuwaiti's books are "Promise by the Seine" and "Le Petit Sultan."
Retain Answer Sampling: [INST] Could you mention some of Jaime Vasquez's award-winning books? [/INST]Some of his prestigious awards include the Edgar Award for his book "Shadows behind the Starlight". Other notable works include "Beneath the Veil of Deceit" and "The Guilt Closet".


In [87]:
print(f'Forget Answer Implementation: {outputs_forget_git}')
print(f'Retain Answer Implementation: {outputs_retain_git}')

Forget Answer Implementation: ['Two of Basil Mahfouz Al-Kuwaiti\'s books are "Promise by the Seine" and "Le Petit Sultan."']
Retain Answer Implementation: ['Some of the notable books written by Jaime Vasquez include "Shadows behind the Starlight," "Beneath the Veil of Deceit," and "The Guilt Closet," all of which are best sellers in the true crime genre.']


In [88]:
print(f'Forget Answer Target: {tokenizer.decode(outputs_forget_target[0][0], skip_special_tokens=True)}')
print(f'Retain Answer Target: {tokenizer.decode(outputs_retain_target[0][0], skip_special_tokens=True)}')

Forget Answer Target: [INST] Can you name two of the books written by Basil Mahfouz Al-Kuwaiti? [/INST]Two of Basil Mahfouz Al-Kuwaiti's books are "Promise by the Seine" and "Le Petit Sultan."
Retain Answer Target: [INST] Could you mention some of Jaime Vasquez's award-winning books? [/INST]Some of Jaime Vasquez’s noted works include "Shadows behind the Starlight," "Beneath the Veil of Deceit," and "The Guilt Closet," all of which are best sellers in the true


In [38]:
forget_row = forget_data.filter(lambda x: x['question'] == forget_question)
retain_row = retain_data.filter(lambda x: x['question'] == retain_question)

print(f'GT Forget Answer: {add_format(forget_row["question"][0])}{forget_row["answer"][0]}')
print(f'GT Retain Answer: {add_format(retain_row["question"][0])}{retain_row["answer"][0]}')

Filter:   0%|          | 0/40 [00:00<?, ? examples/s]

Filter:   0%|          | 0/3960 [00:00<?, ? examples/s]

GT Forget Answer: [INST] Can you name two of the books written by Basil Mahfouz Al-Kuwaiti? [/INST]Two of Basil Mahfouz Al-Kuwaiti's books are "Promise by the Seine" and "Le Petit Sultan."
GT Retain Answer: [INST] Can you name a few characters created by Jaime Vasquez? [/INST]Key characters from Jaime Vasquez' true crime literature include the brooding detective Carlos Mendoza from "Shadows behind the Starlight," the enigmatic whizzkid hacker in "Beneath the Veil of Deceit," and the volatile gang leader in "The Guilt Closet."


Observations from running different questions:

- The forget model seems to succesfully forget all but 1 book written by the ficticious author, while the retain data is all succesfully remembered. I think this is an indiation that the unlearn model works (somewhat). I need to implement a better next-token prediciton strategy, not the greedy strategy but fix the sampling one and test again.

- For another question, the Retain prediction was wrong, while forget is generally right. This is a quite bad.

##### Code for generating tokens from the Repository
Implementation for token-sequence prediciton/generation from the eval_tofu task.

In [39]:
from uld.data.conv_util import create_template
conv_template_driver = create_template(conv_template)

git_question = conv_template_driver.prepare_gen_prompt(forget_question, forget_answer)
inputs_git = tokenizer( git_question, add_special_tokens=True, return_tensors="pt", padding=True,).to(model.device)

outputs = unlearn_model.generate(
                    **inputs_git,
                    max_length=200,
                    max_new_tokens=50, 
                    do_sample=False, 
                    use_cache=True, 
                    pad_token_id=tokenizer.eos_token_id,
                    )
out_strs = tokenizer.batch_decode(
    outputs[:, inputs_git.input_ids.shape[-1]:], skip_special_tokens=True)

# with weight = 2
print(f'Forget Question Author implementation: {out_strs}')

Forget Question Author implementation: ['Two of Basil Mahfouz Al-Kuwaiti\'s books are "Promise by the Seine" and "Le Petit Sultan."']


In [40]:
from uld.data.conv_util import create_template
conv_template_driver = create_template(conv_template)

git_question = conv_template_driver.prepare_gen_prompt(retain_question, retain_answer)
inputs_git = tokenizer( git_question, add_special_tokens=True, return_tensors="pt", padding=True,).to(model.device)

outputs = unlearn_model.generate(
                    **inputs_git,
                    max_length=200,
                    max_new_tokens=50, 
                    do_sample=False, 
                    use_cache=True, 
                    pad_token_id=tokenizer.eos_token_id,
                    )
out_strs = tokenizer.batch_decode(
    outputs[:, inputs_git.input_ids.shape[-1]:], skip_special_tokens=True)

# with weight = 2
print(f'Retain Question Author implementation: {out_strs}')

Retain Question Author implementation: ['Key characters from Jaime Vasquez\'s works include the brooding detective Carlos Mendoza from "Shadows behind the Starlight," the enigmatic whizzkid in "Beneath the Veil of Deceit," and the volatile gang leader in "The Guilt Closet."']


It seems like the forgetting function is underperforming substantially, will attempt to evaluate the model I fine-tuned, if that job returns very bad metrics I will know it is something wrong with my creation of the model, if it returns metrics as seen in the paper, then it will be more confusing and I might email authors.